File: apps/ios/Shared/TonoAnalytics.swift
// TonoAnalytics.swift
// A3: Lightweight, first-party event stream. Events go to /v1/events on
// the Tono backend — no third-party SDK, no heavy initialization, no memory
// cost in the keyboard extension at startup.
//
// A4 PRIVACY CONTRACT (enforced here, not downstream):
//   ✓  Event names, axis enums, risk-level enums, latency_ms, mode strings
//   ✓  Anonymized device ID only
//   ✗  NO message text — never include draft or rewrite text
//   ✗  NO recipient names — only axis labels and boolean flags
//   ✗  NO free-text user input of any kind
//
// Call site rule: properties are computed here and checked in code review.
// If a property ever needs to carry user-visible text, it must be replaced
// with an opaque enum value before the track() call.

import Foundation

// MARK: - Event catalog

public enum AnalyticsEvent: Sendable {
    /// User tapped Coach or Read.
    case coachRequested(mode: String)
    /// Analysis result appeared on screen (mock or real LLM).
    case analysisShown(riskLevel: String, latencyMs: Int, source: String)
    /// User tapped a rewrite chip and it was inserted.
    case rewriteInserted(selectedAxis: String, shownAxes: [String])
    /// Draft text changed after insert — user edited the suggested rewrite.
    case rewriteEditedAfterInsert
    /// User inserted one axis but others were shown (derived on back end).
    case axisRejected(shownAxes: [String], pickedAxis: String)
    /// User inserted a word from the inline suggestion strip.
    case suggestionTapped
    /// Collective improvement signal: content-free session outcome.
    /// Respects the improveTono flag — never sent when user has opted out.
    /// Fields: risk enum, axis enum or nil, mode enum, length BUCKET (not length),
    /// bool flags. NO message text, NO rewrite text, NO recipient identifier.
    case improvementOutcome(
        riskLevel: String,
        axisSelected: String?,
        mode: String,
        msgLenBucket: String,
        rewriteUsed: Bool,
        editAfter: Bool
    )

    public var name: String {
        switch self {
        case .coachRequested:           return "coach_requested"
        case .analysisShown:            return "analysis_shown"
        case .rewriteInserted:          return "rewrite_inserted"
        case .rewriteEditedAfterInsert: return "rewrite_edited_after_insert"
        case .axisRejected:             return "axis_rejected"
        case .suggestionTapped:         return "suggestion_tapped"
        case .improvementOutcome:       return "improvement_outcome"
        }
    }

    // A4: Only permitted properties — strings are enum values, not user content.
    public var properties: [String: Any] {
        switch self {
        case .coachRequested(let mode):
            return ["mode": mode]
        case .analysisShown(let risk, let ms, let source):
            return ["risk_level": risk, "latency_ms": ms, "source": source]
        case .rewriteInserted(let axis, let shown):
            return ["selected_axis": axis, "shown_axes": shown]
        case .rewriteEditedAfterInsert:
            return [:]
        case .axisRejected(let shown, let picked):
            return ["shown_axes": shown, "picked_axis": picked]
        case .suggestionTapped:
            return [:]
        case .improvementOutcome(let risk, let axis, let mode, let bucket, let used, let edit):
            var props: [String: Any] = [
                "risk_level": risk,
                "mode": mode,
                "msg_len_bucket": bucket,
                "rewrite_used": used,
                "edit_after": edit,
            ]
            if let axis { props["selected_axis"] = axis }
            return props
        }
    }
}

// MARK: - Client

public enum TonoAnalytics {
    /// Fire-and-forget: failures are silently discarded.
    /// Never blocks the calling thread; dispatches on a background URLSession.
    public static func track(_ event: AnalyticsEvent) {
        // Collective improvement events respect the opt-out flag — never sent
        // when the user has disabled "Help improve Tono" in Settings.
        if case .improvementOutcome = event, !FeatureFlags.isEnabled(.improveTono) { return }

        let deviceId = SharedKeychain.get(KeychainKeys.deviceID) ?? ""
        guard !deviceId.isEmpty else { return }
        let token = SharedKeychain.get(KeychainKeys.apiToken) ?? ""

        var body: [String: Any] = [
            "event": event.name,
            "device_id": deviceId,
            "ts": Int(Date().timeIntervalSince1970),
        ]
        event.properties.forEach { body[$0.key] = $0.value }

        guard let data = try? JSONSerialization.data(withJSONObject: body) else { return }
        let url = TonoBackend.shared.baseURL.appendingPathComponent("v1/events")
        var req = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 8)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if !token.isEmpty {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        req.httpBody = data

        // Fire-and-forget on the default session (no response handling).
        URLSession.shared.dataTask(with: req).resume()
    }
}

File: apps/android/shared/src/main/java/com/tono/shared/analytics/TonoAnalytics.kt
package com.tono.shared.analytics

import com.tono.shared.flags.FeatureFlag
import com.tono.shared.flags.FeatureFlags
import com.tono.shared.network.TonoBackend
import com.tono.shared.storage.KeychainKeys
import com.tono.shared.storage.SecureStore
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException

// Mirrors ios/Shared/TonoAnalytics.swift
//
// A4 PRIVACY CONTRACT (enforced here, not downstream):
//   ✓  Event names, axis enums, risk-level enums, latency_ms, mode strings
//   ✓  Anonymized device ID only
//   ✗  NO message text — never include draft or rewrite text
//   ✗  NO recipient names — only axis labels and boolean flags
//   ✗  NO free-text user input of any kind

sealed class AnalyticsEvent {
    data class CoachRequested(val mode: String) : AnalyticsEvent()
    data class AnalysisShown(val riskLevel: String, val latencyMs: Int, val source: String) : AnalyticsEvent()
    data class RewriteInserted(val selectedAxis: String, val shownAxes: List<String>) : AnalyticsEvent()
    object RewriteEditedAfterInsert : AnalyticsEvent()
    data class AxisRejected(val shownAxes: List<String>, val pickedAxis: String) : AnalyticsEvent()
    // Collective improvement: content-free session outcome.
    // Respects IMPROVE_TONO flag — never sent when user has opted out.
    // NO message text, NO rewrite text, NO recipient identifier.
    data class ImprovementOutcome(
        val riskLevel: String,
        val axisSelected: String?,
        val mode: String,
        val msgLenBucket: String,
        val rewriteUsed: Boolean,
        val editAfter: Boolean,
    ) : AnalyticsEvent()

    val name: String get() = when (this) {
        is CoachRequested           -> "coach_requested"
        is AnalysisShown            -> "analysis_shown"
        is RewriteInserted          -> "rewrite_inserted"
        is RewriteEditedAfterInsert -> "rewrite_edited_after_insert"
        is AxisRejected             -> "axis_rejected"
        is ImprovementOutcome       -> "improvement_outcome"
    }

    // A4: only permitted properties — no user content
    val properties: Map<String, Any> get() = when (this) {
        is CoachRequested           -> mapOf("mode" to mode)
        is AnalysisShown            -> mapOf("risk_level" to riskLevel, "latency_ms" to latencyMs, "source" to source)
        is RewriteInserted          -> mapOf("selected_axis" to selectedAxis, "shown_axes" to shownAxes)
        is RewriteEditedAfterInsert -> emptyMap()
        is AxisRejected             -> mapOf("shown_axes" to shownAxes, "picked_axis" to pickedAxis)
        is ImprovementOutcome       -> buildMap {
            put("risk_level", riskLevel)
            put("mode", mode)
            put("msg_len_bucket", msgLenBucket)
            put("rewrite_used", rewriteUsed)
            put("edit_after", editAfter)
            axisSelected?.let { put("selected_axis", it) }
        }
    }
}

object TonoAnalytics {
    private val client = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    fun track(event: AnalyticsEvent) {
        // Collective improvement events respect the opt-out flag.
        if (event is AnalyticsEvent.ImprovementOutcome &&
            !FeatureFlags.isEnabled(FeatureFlag.IMPROVE_TONO)) return

        val deviceId = SecureStore.get(KeychainKeys.DEVICE_ID)?.takeIf { it.isNotBlank() } ?: return
        val token = SecureStore.get(KeychainKeys.API_TOKEN) ?: ""

        val body: MutableMap<String, Any> = mutableMapOf(
            "event" to event.name,
            "device_id" to deviceId,
            "ts" to (System.currentTimeMillis() / 1000).toInt(),
        )
        body.putAll(event.properties)

        val bodyJson = json.encodeToString(
            kotlinx.serialization.json.buildJsonObject {
                body.forEach { (k, v) ->
                    when (v) {
                        is String -> put(k, kotlinx.serialization.json.JsonPrimitive(v))
                        is Int    -> put(k, kotlinx.serialization.json.JsonPrimitive(v))
                        is Boolean -> put(k, kotlinx.serialization.json.JsonPrimitive(v))
                        is List<*> -> put(k, kotlinx.serialization.json.buildJsonArray {
                            v.forEach { item -> if (item is String) add(kotlinx.serialization.json.JsonPrimitive(item)) }
                        })
                        else -> {}
                    }
                }
            }
        )

        val request = Request.Builder()
            .url("${TonoBackend.baseUrl}/v1/events")
            .post(bodyJson.toRequestBody("application/json".toMediaType()))
            .header("Content-Type", "application/json")
            .apply { if (token.isNotBlank()) header("Authorization", "Bearer $token") }
            .build()

        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) = Unit
            override fun onResponse(call: Call, response: Response) { response.close() }
        })
    }
}
